### PR TITLE
local_ratelimit: allow sharing token bucket between filter chains

### PR DESCRIPTION
**Pull request by ggreenway**
Commit Message: local_ratelimit: allow sharing token bucket between filter chains
Additional Description:
Risk Level: Low
Testing: New unit and integration tests
Docs Changes: updated in protos
Release Notes: added
Platform Specific Features: none
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/extensions/filters/network/local_ratelimit/local_ratelimit_integration_test.cc [hunk 1]
```diff
@@ -46,6 +48,83 @@ name: ratelimit
...
+  "@type": type.googleapis.com/envoy.extensions.filters.network.local_ratelimit.v3.LocalRateLimit
+  stat_prefix: local_rate_limit_stats
+  share_key: 'the_key'
+  token_bucket:
+    max_tokens: 2
+    # Set fill_interval to effectively infinite so we only get max_tokens to start and never re-fill.
+    fill_interval: 100000s
+)EOF");
+
+    // Clone the whole listener, which includes the `share_key`.
```
Comment (daixiang0) [inline test/extensions/filters/network/local_ratelimit/local_ratelimit_integration_test.cc:94]: Clone is not what the end-user does, is it better to add a new filter config to test?

File: api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto [hunk 2]
```diff
@@ -44,4 +45,9 @@ message LocalRateLimit {
   // Runtime flag that controls whether the filter is enabled or not. If not specified, defaults
   // to enabled.
   config.core.v3.RuntimeFeatureFlag runtime_enabled = 3;
+
+  // Specifies that the token bucket used for rate limiting should be shared with other local_rate_limit filters
+  // with a matching :ref:`token_bucket <envoy_v3_api_field_extensions.filters.network.local_ratelimit.v3.LocalRateLimit.token_bucket>`
+  // and `share_key` configuration. All fields of `token_bucket` must match exactly for the token bucket to be shared.
+  string share_key = 4;
```
Comment (adisuissa) [inline api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto:52]: Will this be used also for http filters?
I'm trying to understand if this should be part of the `TokenBucket` type, as it is highly coupled with it (i.e., for filters to share this there needs to be equality of all fields in the bucket and this field). In other words, is this a property of the filter or the bucket?

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.cc [hunk 3]
```diff
@@ -10,26 +10,60 @@ namespace Extensions {
...
+    return create_fn();
+  }
+
+  const Key key{proto_config.share_key(), proto_config.token_bucket()};
+  auto it = limiters_.find(key);
+  if (it == limiters_.end()) {
+    ENVOY_LOG(debug, fmt::format("LocalRateLimit for share_key '{}' is creating a new token bucket "
+                                 "due to no match found",
+                                 proto_config.share_key()));
+    it = limiters_.emplace(key, create_fn()).first;
```
Comment (adisuissa) [inline source/extensions/filters/network/local_ratelimit/local_ratelimit.cc:29]: I might be missing something here, but isn't this singleton shared across all worker threads?
If so, couldn't there be a race here where multiple emplace calls will be done with the same key?

**Review by adisuissa — COMMENTED** [review]
Thanks for adding this feature!
Left a couple of questions.

File: test/extensions/filters/network/local_ratelimit/local_ratelimit_integration_test.cc (hunk 1 above)
Comment (ggreenway) [reply test/extensions/filters/network/local_ratelimit/local_ratelimit_integration_test.cc:94]: This is just an easy way to manipulate the protobufs in the test. It yields the same result as copy/paste the yaml to get the same fields.

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.cc (hunk 3 above)
Comment (ggreenway) [reply source/extensions/filters/network/local_ratelimit/local_ratelimit.cc:29]: This code all runs at config-load time, which is always on the main thread. Adding `ASSERT_IS_MAIN_OR_TEST_THREAD` to make it clear and validate that it is true.

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.cc [hunk 4]
```diff
@@ -10,26 +10,62 @@ namespace Extensions {
...
-          dispatcher,
-          Protobuf::RepeatedPtrField<
-              envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>())),
-      enabled_(proto_config.runtime_enabled(), runtime),
-      stats_(generateStats(proto_config.stat_prefix(), scope)) {}
+    Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime,
+    Singleton::Manager& singleton_manager)
+    : enabled_(proto_config.runtime_enabled(), runtime),
+      stats_(generateStats(proto_config.stat_prefix(), scope)),
+      shared_bucket_registry_(singleton_manager.getTyped<SharedRateLimitSingleton>(
```
Comment (zhxie) [inline source/extensions/filters/network/local_ratelimit/local_ratelimit.cc:47]: What will happen if the config get updated (e.g., by xDS) ? Will it reuse the singleton before update?

Comment (ggreenway) [reply source/extensions/filters/network/local_ratelimit/local_ratelimit.cc:67]: The `Singleton::Manager` is owned by the `Server`, so it's lifetime is the same as the process, and objects are held by `weak_ptr`. So as long as a `local_ratelimit` filter exists, this singleton instance will exist. The token buckets in this singleton instance are held by `shared_ptr`, so they exist as long as at least 1 filter is still using them.

If the configuration is updated for this filter (eg the rate configuration is changed, or the `share_key` is changed), a new instance will be created for that configuration.

So in summary, token buckets will continue to exist as long as they are still in use.

But this question made me realize that the singleton will never free old, unused instances from `limiters_`. I pushed a fix for that issue.

File: api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto (hunk 2 above)
Comment (mattklein123) [reply api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto:52]: Yeah I would be concerned about use cases for TokenBucket that don't share so it seems OK to put it here, though I don't feel strongly about it. I could go either way.

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.h [hunk 5]
```diff
@@ -27,28 +28,66 @@ struct LocalRateLimitStats {
...
-  void onFillTimer();
 
-  Filters::Common::LocalRateLimit::LocalRateLimiterImpl rate_limiter_;
+  LocalRateLimiterImplSharedPtr rate_limiter_;
   Runtime::FeatureFlag enabled_;
   LocalRateLimitStats stats_;
+  const SharedRateLimitSingleton::Key* shared_bucket_key_{};
+  std::shared_ptr<SharedRateLimitSingleton> shared_bucket_registry_;
 
-  std::vector<RateLimit::LocalDescriptor> descriptors_;
```
Comment (mattklein123) [inline source/extensions/filters/network/local_ratelimit/local_ratelimit.h:51]: This was previously unused, right? Was it there just to avoid a temporary allocation? If so you might want to leave it and make it const and add a small comment? Maybe I'm reading the diff wrong.

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.cc [hunk 6]
```diff
@@ -10,26 +10,90 @@ namespace Extensions {
...
+        debug,
+        fmt::format("LocalRateLimit for share_key '{}' is using an existing matching token bucket",
+                    proto_config.share_key()));
+    rate_limiter = it->second.lock();
+    ASSERT(rate_limiter != nullptr);
+  }
+  return {rate_limiter, &it->first};
+}
+
+void SharedRateLimitSingleton::removeIfUnused(const Key* key) {
```
Comment (mattklein123) [inline source/extensions/filters/network/local_ratelimit/local_ratelimit.cc:51]: Potentially add `ASSERT_IS_MAIN_OR_TEST_THREAD();` here also? I know we have had lots of bugs over the years with things accidentally getting run on other threads, and this would not be thread safe if so.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comments.

/wait

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.h (hunk 5 above)
Comment (ggreenway) [reply source/extensions/filters/network/local_ratelimit/local_ratelimit.h:51]: It was always empty. But the function it is passed to takes an `absl::Span` not a `std::vector&`, so we don't need to construct a vector at all anymore; just pass an empty Span.

**Review by mattklein123 — APPROVED** [review]
Thanks!

File: api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto (hunk 2 above)
Comment (adisuissa) [reply api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto:52]: Ok, I think it makes sense.
I suggest adding to the comment what happens if share_key isn't set (e.g., If no share_key is provided each filter will have its own rate limit).

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.cc [hunk 7]
```diff
@@ -10,26 +10,91 @@ namespace Extensions {
...
+
+  LocalRateLimiterImplSharedPtr rate_limiter;
+  const Key key{proto_config.share_key(), proto_config.token_bucket()};
+  auto it = limiters_.find(key);
+  if (it == limiters_.end()) {
+    ENVOY_LOG(debug, fmt::format("LocalRateLimit for share_key '{}' is creating a new token bucket "
+                                 "due to no match found",
+                                 proto_config.share_key()));
+    rate_limiter = create_fn();
+    it = limiters_.emplace(key, rate_limiter).first;
```
Comment (adisuissa) [inline source/extensions/filters/network/local_ratelimit/local_ratelimit.cc:39]: Following up on our previous discussion here (whether it is thread safe, and that this is part of the config):
Isn't this singleton table accessed by the worker-threads as it is part of the shared config?
I guess the naming "Config" is a bit confusing me, especially the comment [here](https://github.com/envoyproxy/envoy/blob/cefb8f866e393ed87f54c6c030b5b82d9c11037a/source/extensions/filters/network/local_ratelimit/local_ratelimit.h#L60-L63).

I guess that the assertion above holds if the main thread is always the one that creates this Config and just passes it along to other threads, so as long as the `shared_bucket_registry_` is only accessed in the c'tor (or by methods that aren't called by worker threads) then it should be ok.

**Review by adisuissa — COMMENTED** [review]
Thanks for the clarifications.
Overall LGTM, and agree with the API.
Let me know if you think the API comment suggestion is not needed.

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.cc (hunk 7 above)
Comment (ggreenway) [reply source/extensions/filters/network/local_ratelimit/local_ratelimit.cc:39]: Yep, it works because the Config constructor does the lookup, and it is always constructed on the main thread.

File: api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto (hunk 2 above)
Comment (ggreenway) [reply api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto:53]: I documented that empty means no sharing.

File: source/extensions/filters/network/local_ratelimit/local_ratelimit.cc (hunk 7 above)
Comment (adisuissa) [reply source/extensions/filters/network/local_ratelimit/local_ratelimit.cc:39]: If that’s the case then I think that adding the assertion should ensure that it won’t be misused.

**Review by adisuissa — APPROVED** [review]
/lgtm api
LGTM code changes, thanks!
